tools/urinfo/urinfo.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2023 Intel Corporation
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+// See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "urinfo.hpp"
+#include <cstdlib>
+#include <iostream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <vector>
+
+namespace urinfo {
+struct app {
+    bool verbose = false;
+    ur_loader_config_handle_t loaderConfig = nullptr;
+    std::vector<ur_adapter_handle_t> adapters;
+    std::unordered_map<ur_adapter_handle_t, std::vector<ur_platform_handle_t>>
+        adapterPlatformsMap;
+    std::unordered_map<ur_platform_handle_t, std::vector<ur_device_handle_t>>
+        platformDevicesMap;
+
+    app(int argc, const char **argv) {
+        parseArgs(argc, argv);
+        UR_CHECK(urLoaderConfigCreate(&loaderConfig));
+        UR_CHECK(urLoaderConfigEnableLayer(loaderConfig,
+                                           "UR_LAYER_FULL_VALIDATION"));
+        UR_CHECK(urInit(0, loaderConfig));
+        enumerateDevices();
+    }
+
+    void parseArgs(int argc, const char **argv) {
+        static const char *usage = R"(usage: %s [-h] [-v] [-V]
+
+This tool enumerates Unified Runtime layers, adapters, platforms, and
+devices which are currently visible in the local execution environment.
+
+options:
+  -h, --help            show this help message and exit
+  --version             show version number and exit
+  -v, --verbose         print additional information
+)";
+        for (int argi = 1; argi < argc; argi++) {
+            std::string_view arg{argv[argi]};
+            if (arg == "-h" || arg == "--help") {
+                std::printf(usage, argv[0]);
+                std::exit(0);
+            } else if (arg == "--version") {
+                std::printf("%s v%s\n", argv[0], UR_VERSION);
+                std::exit(0);
+            } else if (arg == "-v" || arg == "--verbose") {
+                verbose = true;
+            } else {
+                std::fprintf(stderr, "error: invalid argument: %s\n",
+                             argv[argi]);
+                std::fprintf(stderr, usage, argv[0]);
+                std::exit(1);
+            }
+        }
+    }
+
+    void enumerateDevices() {
+        // Enumerate adapters.
+        uint32_t numAdapters = 0;
+        UR_CHECK(urAdapterGet(0, nullptr, &numAdapters));
+        if (numAdapters == 0) {
+            std::cout << "No adapters found.\n";
+            std::exit(0);
+        }
+        adapters.resize(numAdapters);
+        UR_CHECK(urAdapterGet(numAdapters, adapters.data(), nullptr));
+
+        for (auto adapter : adapters) {
+            // Enumerate platforms
+            uint32_t numPlatforms = 0;
+            UR_CHECK(urPlatformGet(&adapter, 1, 0, nullptr, &numPlatforms));
+            if (numPlatforms == 0) {
+                std::cout << "No platforms found.\n";
+                std::exit(0);
```

Comment:
`continue` ? otherwise this will exit after the first adapter with no platforms.

Reply:
Indeed, should do the same for devices in platforms too.